architecture/design/pessimistic-locking-functional-spec.md

```diff
@@ -0,0 +1,661 @@
+
+<span style="text-decoration:underline;">Conflicts in YSQL -</span>
+
+YSQL does not differentiate between a lock-lock conflict and a lock-modification conflict. In either case, we will trigger conflict resolution. Conflict resolution in YSQL doesn’t lead to any blocking, but would surely abort either itself or all conflicting transactions based on priorities of the transactions (which are randomly chosen). This is _optimistic locking_.
+
+Moreover, unlike Pg’s Serializable isolation, in YSQL’s Serializable isolation, plain SELECTs without explicit lock also implicitly take _FOR SHARE_ lock. And this can result in a plain read DML to conflict with a write DML (which will result in triggering conflict resolution).
+
+<span style="text-decoration:underline;">Pessimistic Locking in PostgreSQL -</span>
+
+PostgreSQL uses **_pessimistic locking_**, where a transaction will **wait upon encountering a lock-lock conflict** in order to acquire the locks it needs and proceed.
+
+**Pessimistic locking (In Pg): **When a txn T1 attempts to take a lock (either implicitly via a write or explicitly using `SELECT FOR`), it might be blocked by txns that hold a conflicting lock (note again, either implicitly via a write or explicitly using `SELECT FOR`). The txn will then wait for all lock-lock conflicting transactions to end before obtaining the lock. Once the other txns end, either the lock will be taken or a **lock-modification** conflict will be detected and a serialization error will occur.
+
+Some points to note are -
+
+
+
+1. **Txn might skip being added to the queue:** If a transaction is trying to acquire a lock that doesn’t conflict with actively held locks but does conflict with a transaction that is already waiting to acquire a conflicting lock, it proceeds to take the lock. In short, the transaction skips being added to the queue itself (refer example 3). This could also lead to starvation.
+2. **Txn already in queue might jump older txns in the queue: **Multiple transactions might be blocked to lock (either implicitly via write or explicitly) the same tuple and this results in a queue. When an active txn ends, txns that entered the queue earlier are checked first for unblocking. Txns that don’t conflict with the active txns are unblocked to acquire locks and become active. Note that this implies that a transaction that conflicts with earlier transactions blocked in the queue might still be unblocked if it doesn’t conflict with active transactions (which could lead to starvation).
+3. In case <code>[statement_timeout or lock_timeout](https://www.postgresql.org/docs/13/runtime-config-client.html)</code> is non-zero, a blocked txn will abort after the timeout. (example 4). A zero timeout would imply indefinite waiting.
+
+NOTE: Refer this [link](https://postgrespro.com/blog/pgsql/5968005) for a good overview of row-level locking.
```

Comment:
This article, this blog post... anything but link

Reply:
done